pkg/registry/generic/rest/proxy.go

```diff
@@ -192,6 +209,104 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
 	return true
 }
 
+// connectBackend dials the backend at location and forwards a copy of the client request.
+func (h *UpgradeAwareProxyHandler) connectBackend(req *http.Request, location *url.URL) (conn net.Conn, err error) {
+	defer func() {
+		if err != nil && conn != nil {
+			conn.Close()
+		}
+	}()
+
+	beReq, err := http.NewRequest(req.Method, location.String(), req.Body)
+	if err != nil {
+		return nil, err
+	}
+	beReq.Header = req.Header
+
+	conn, err = proxy.DialURL(location, h.Transport)
+	if err != nil {
+		return conn, fmt.Errorf("error dialing backend: %v", err)
+	}
+
+	if err = beReq.Write(conn); err != nil {
+		return conn, fmt.Errorf("error sending request: %v", err)
+	}
+
+	return conn, err
+}
+
+// connectBackendWithRedirects dials the backend and forwards a copy of the client request. If the
+// client responds with a redirect, it is followed. The raw response bytes are returned, and should
+// be forwarded back to the client.
+func (h *UpgradeAwareProxyHandler) connectBackendWithRedirects(req *http.Request) (net.Conn, []byte, error) {
+	const (
+		maxRedirects    = 10
+		maxResponseSize = 1024
+	)
+	var (
+		rawResponse      = bytes.NewBuffer(make([]byte, 0, 256))
+		location         = h.Location
+		intermediateConn net.Conn
+		err              error
+	)
+	defer func() {
+		if intermediateConn != nil {
+			intermediateConn.Close()
+		}
+	}()
+
+redirectLoop:
+	for redirects := 0; ; redirects++ {
+		if redirects == maxRedirects {
+			return nil, nil, fmt.Errorf("too many redirects (%d)", redirects)
+		}
+
+		intermediateConn, err = h.connectBackend(req, location)
```

Comment:
Thinking about this a bit more... If we decide to move to HTTP/2 for streaming requests, and if there are no changes to go's HTTP/2 library, then the only means we'll have of streaming data over the wire will be via the request and response bodies (we'd need to implementing muxing on top). Which would mean that when the client has data it wants to send to the server, the original request body needs to be preserved and available. I'm not sure how that would work... Also not sure it needs to stop this from going in.